core/src/com/unciv/logic/multiplayer/chat/ChatWebSocketManager.kt

```diff
@@ -0,0 +1,249 @@
+import kotlinx.serialization.json.ClassDiscriminatorMode
+import kotlinx.serialization.json.Json
+
+// used when sending a message
+@Serializable
+sealed class Message {
+    @Serializable
+    @SerialName("chat")
+    data class Chat(
+        val civName: String, val gameId: String, val message: String
+    ) : Message()
+
+    @Serializable
+    @SerialName("join")
+    data class Join(
+        val gameIds: List<String>
+    ) : Message()
+
+    @Serializable
+    @SerialName("leave")
+    data class Leave(
+        val gameIds: List<String>
+    ) : Message()
+}
+
+// used when receiving a message
+@Serializable
+sealed class Response {
+    @Serializable
+    @SerialName("chat")
+    data class Chat(
+        val civName: String, val gameId: String, val message: String
+    ) : Response()
+
+    @Serializable
+    @SerialName("joinSuccess")
+    data class JoinSuccess(
+        val gameIds: List<String>
+    ) : Response()
+
+    @Serializable
+    @SerialName("error")
+    data class Error(
+        val message: String
+    ) : Response()
+}
+
+
+object ChatWebSocketManager {
+    private var errorReconnectionAttempts = 0
+    private const val MAX_ERROR_RECONNECTION_ATTEMPTS = 100
+    private const val RECONNECT_TIME_MS: Long = 5_000
+
+    private var job: Job? = null
+    private val eventReceiver = EventBus.EventReceiver()
+    private var session: DefaultClientWebSocketSession? = null
+
+    @OptIn(ExperimentalSerializationApi::class)
+    private val client = HttpClient(CIO) {
+        install(WebSockets) {
+            pingInterval = 30_000
+            contentConverter = KotlinxWebsocketSerializationConverter(Json {
+                classDiscriminator = "type"
+                // DO NOT OMIT
+                // if omitted the "type" field will be missing from all outgoing messages
+                classDiscriminatorMode = ClassDiscriminatorMode.ALL_JSON_OBJECTS
+            })
+        }
+    }
+
+    private val chatUrl
+        get() = URLBuilder(
+            UncivGame.Current.onlineMultiplayer.multiplayerServer.getServerUrl()
+        ).apply {
+            appendPathSegments("chat")
+            protocol = if (protocol.isSecure()) URLProtocol.WSS else URLProtocol.WS
+        }.build()
+
+    /*
+     * Delivery not guaranted & failures are mosly ignored.
+     */
+    fun requestMessageSend(message: Message) {
+        startSocket()
+        Concurrency.run("MultiplayerChatSendMessage") {
+            withTimeoutOrNull(1000) {
+                while (session == null) {
+                    delay(50)
+                    yield()
+                }
+            }
+            session?.runCatching {
+                this.sendSerialized(message)
+            }
+        }
+    }
+
+    fun handleWebSocketThrowables(t: Throwable) {
+        println("ChatError: ${t.message}")
+
+        if (errorReconnectionAttempts == 0)
+            relayGlobalMessage("WebSocket connection closed. Cause: ${t.cause}!")
+
+        if (
```

Comment:
Else if?

Reply:
No, we only want send global message the first time websocket crashes. So, this is as intended.

Reply:
Oh sorry. fixed by https://github.com/yairm210/Unciv/pull/13599/commits/95c8dc6511719a16f7dd8763384395996d363e6a